vcd/resource_vcd_vapp_nat_rules.go

```diff
@@ -0,0 +1,306 @@
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+	vcdClient.lockParentVappWithName(d, vapp.VApp.Name)
+	defer vcdClient.unLockParentVappWithName(d, vapp.VApp.Name)
+
+	networkId := d.Get("network_id").(string)
+	natType := d.Get("nat_type").(string)
+	netRules, err := expandVappNetworkNatRules(d, vapp, natType)
+	if err != nil {
+		return fmt.Errorf("error expanding NAT rules: %s", err)
+	}
+	policy := allowTrafficInPolicy
+	if !d.Get("enable_ip_masquerade").(bool) && natType == portForwardingNatType {
+		policy = allowTrafficPolicy
+	}
+	vappNetwork, err := vapp.UpdateNetworkNatRules(networkId, netRules,
+		natType, policy)
+	if err != nil {
+		log.Printf("[INFO] Error setting NAT rules: %s", err)
+		return fmt.Errorf("error setting NAT rules: %s", err)
+	}
+
+	d.SetId(vappNetwork.ID)
+
+	return resourceVappNetworkNatRulesRead(d, meta)
+}
+
+func resourceVAppNetworkNatRulesDelete(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vcdClient.lockParentVappWithName(d, vapp.VApp.Name)
+	defer vcdClient.unLockParentVappWithName(d, vapp.VApp.Name)
+
+	err = vapp.RemoveAllNetworkNatRules(d.Get("network_id").(string))
+	if err != nil {
+		log.Printf("[INFO] Error deleting NAT rules: %s", err)
+		return fmt.Errorf("error deleting NAT rules: %s", err)
+	}
+
+	return nil
+}
+
+func resourceVappNetworkNatRulesRead(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vappNetwork, err := vapp.GetVappNetworkById(d.Get("network_id").(string), false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp network. %s", err)
+	}
+
+	var rules []map[string]interface{}
+	if vappNetwork.Configuration.Features == nil || vappNetwork.Configuration.Features.NatService == nil {
+		log.Print("no Nat rules found.")
+		_ = d.Set("rule", rules)
+	}
+
+	for _, rule := range vappNetwork.Configuration.Features.NatService.NatRule {
+		singleRule := make(map[string]interface{})
+		singleRule["id"] = rule.ID
+		if vappNetwork.Configuration.Features.NatService.NatType == portForwardingNatType {
```

Comment:
Please consider using a `switch` statement instead of this cumbersome `if`

Reply:
changed, though I am not sure is better, maybe